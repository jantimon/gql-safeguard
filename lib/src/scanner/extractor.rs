use anyhow::Result;
use std::fs;
use std::path::Path;
use swc_core::ecma::{
    ast::*,
    parser::{lexer::Lexer, Parser, StringInput, Syntax, TsConfig},
    visit::{Visit, VisitWith},
};

pub struct GraphQLExtractor;

impl GraphQLExtractor {
    pub fn new() -> Self {
        Self
    }

    pub fn extract_from_file(&self, file_path: &Path) -> Result<Vec<GraphQLString>> {
        let source_code = fs::read_to_string(file_path)?;
        let syntax = if file_path.extension().and_then(|s| s.to_str()) == Some("tsx") {
            Syntax::Typescript(TsConfig {
                tsx: true,
                ..Default::default()
            })
        } else {
            Syntax::Typescript(TsConfig {
                tsx: false,
                ..Default::default()
            })
        };

        let lexer = Lexer::new(
            syntax,
            Default::default(),
            StringInput::new(&source_code, Default::default(), Default::default()),
            None,
        );

        let mut parser = Parser::new_from(lexer);
        let module = parser
            .parse_module()
            .map_err(|e| anyhow::anyhow!("Parse error: {:?}", e))?;

        let mut visitor = GraphQLVisitor::new(file_path.to_path_buf());
        module.visit_with(&mut visitor);

        Ok(visitor.graphql_strings)
    }
}

#[derive(Debug, Clone)]
pub struct GraphQLString {
    pub content: String,
    pub file_path: std::path::PathBuf,
    pub line: usize,
}

struct GraphQLVisitor {
    file_path: std::path::PathBuf,
    graphql_strings: Vec<GraphQLString>,
}

impl GraphQLVisitor {
    fn new(file_path: std::path::PathBuf) -> Self {
        Self {
            file_path,
            graphql_strings: Vec::new(),
        }
    }

    fn extract_graphql_from_tagged_template(&mut self, tpl: &TaggedTpl) {
        if let Expr::Ident(ident) = &*tpl.tag {
            if ident.sym.as_ref() == "gql" || ident.sym.as_ref() == "graphql" {
                if let Some(first_quasi) = tpl.tpl.quasis.first() {
                    let content = first_quasi.raw.as_ref();
                    // Simple line counting - could be improved
                    let line = 1; // Placeholder - would need proper source map

                    self.graphql_strings.push(GraphQLString {
                        content: content.to_string(),
                        file_path: self.file_path.clone(),
                        line,
                    });
                }
            }
        }
    }
}

impl Visit for GraphQLVisitor {
    fn visit_tagged_tpl(&mut self, tpl: &TaggedTpl) {
        self.extract_graphql_from_tagged_template(tpl);
        tpl.visit_children_with(self);
    }
}
