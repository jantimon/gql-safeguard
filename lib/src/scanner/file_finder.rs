use anyhow::Result;
use globset::{Glob, GlobSet, GlobSetBuilder};
use std::path::{Path, PathBuf};
use walkdir::WalkDir;

pub struct FileFinder {
    glob_set: GlobSet,
}

impl FileFinder {
    pub fn new(pattern: &str) -> Result<Self> {
        let mut builder = GlobSetBuilder::new();
        builder.add(Glob::new(pattern)?);
        let glob_set = builder.build()?;
        Ok(Self { glob_set })
    }

    pub fn find_files(&self, root_path: &Path) -> Result<Vec<PathBuf>> {
        let mut matching_files = Vec::new();

        for entry in WalkDir::new(root_path)
            .follow_links(false)
            .into_iter()
            .filter_map(|e| e.ok())
        {
            let path = entry.path();
            if path.is_file() {
                if let Some(path_str) = path.to_str() {
                    if self.glob_set.is_match(path_str) {
                        matching_files.push(path.to_path_buf());
                    }
                }
            }
        }

        Ok(matching_files)
    }
}
