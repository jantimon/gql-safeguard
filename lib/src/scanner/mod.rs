pub mod extractor;
pub mod file_finder;
