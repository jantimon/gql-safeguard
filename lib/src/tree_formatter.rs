pub struct TreeFormatter {
    lines: Vec<(usize, String)>,
    max_depth: usize,
}

impl Default for TreeFormatter {
    fn default() -> Self {
        Self::new()
    }
}

impl std::fmt::Display for TreeFormatter {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.format_tree())
    }
}

impl TreeFormatter {
    pub fn new() -> Self {
        Self {
            lines: Vec::new(),
            max_depth: 0,
        }
    }

    /// Adds a line to the tree structure
    ///
    /// # Arguments
    /// * `depth` - The depth level (0 = root, 1 = first level, etc.)
    /// * `text` - The text content to display
    pub fn add_line(&mut self, depth: usize, text: &str) {
        self.lines.push((depth, text.to_string()));
        self.max_depth = self.max_depth.max(depth);
    }

    /// Adds an entire TreeFormatter as a subtree at the specified depth
    ///
    /// # Arguments
    /// * `depth` - The depth level to add the subtree at
    /// * `tree` - The TreeFormatter to add as a subtree
    pub fn add_tree(&mut self, depth: usize, tree: &TreeFormatter) {
        for (line_depth, text) in &tree.lines {
            let new_depth = depth + line_depth;
            self.lines.push((new_depth, text.clone()));
            self.max_depth = self.max_depth.max(new_depth);
        }
    }

    /// Formats the entire tree structure as a string
    fn format_tree(&self) -> String {
        let mut result = String::new();
        let max_level = self.max_depth + 1;
        let mut last_at_levels = vec![vec![false; self.lines.len()]; max_level];
        for (i, (depth, text)) in self.lines.iter().enumerate() {
            let is_last = self.is_last_sibling(i, *depth);
            last_at_levels[*depth][i] = is_last;
            let formatted_line = self.format_line_internal(*depth, i, text, &last_at_levels);
            result.push_str(&formatted_line);
            if i < self.lines.len() - 1 {
                result.push('\n');
            }
        }

        result
    }

    /// Helper to determine if a line is the last sibling at its level
    fn is_last_sibling(&self, line_index: usize, level: usize) -> bool {
        // Look for the next line at the same level or shallower
        for i in (line_index + 1)..self.lines.len() {
            let next_depth = self.lines[i].0;
            if next_depth < level {
                // Found a shallower line, so this was the last at this level
                return true;
            }
            if next_depth == level {
                // Found another line at the same level, so this is not the last
                return false;
            }
        }
        // No more lines at this level or shallower, so this is the last
        true
    }

    /// Internal method to format a single line using pre-computed last_at_levels
    fn format_line_internal(
        &self,
        depth: usize,
        line_index: usize,
        text: &str,
        last_at_levels: &[Vec<bool>],
    ) -> String {
        let mut result = String::new();

        // Build the indentation prefix for parent levels (depth - 1)
        for level in 1..depth {
            // Check if any ancestor at this level is the last sibling
            let mut use_empty_space = false;

            // Find the most recent ancestor at this level
            for i in (0..=line_index).rev() {
                if self.lines[i].0 == level {
                    if level < last_at_levels.len() && i < last_at_levels[level].len() {
                        use_empty_space = last_at_levels[level][i];
                    }
                    break;
                }
            }

            if use_empty_space {
                result.push_str("    "); // Empty space for completed branches
            } else {
                result.push_str("|   "); // Continuing branch
            }
        }

        // Add the tree symbol for the current level
        if depth > 0 && depth < last_at_levels.len() && line_index < last_at_levels[depth].len() {
            if last_at_levels[depth][line_index] {
                result.push_str("└── ");
            } else {
                result.push_str("├── ");
            }
        }

        result.push_str(text);
        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_single_root_item() {
        let mut formatter = TreeFormatter::new();
        formatter.add_line(0, "root item");
        let result = formatter.to_string();
        assert_eq!(result, "root item");
    }

    #[test]
    fn test_simple_tree() {
        let mut formatter = TreeFormatter::new();
        formatter.add_line(0, "root");
        formatter.add_line(1, "child1");
        formatter.add_line(1, "child2");
        let result = formatter.to_string();
        assert_eq!(result, "root\n├── child1\n└── child2");
    }

    #[test]
    fn test_nested_tree() {
        let mut formatter = TreeFormatter::new();
        formatter.add_line(0, "root");
        formatter.add_line(1, "child1");
        formatter.add_line(2, "grandchild1");
        formatter.add_line(2, "grandchild2");
        formatter.add_line(1, "child2");
        let result = formatter.to_string();
        assert_eq!(
            result,
            "root\n├── child1\n|   ├── grandchild1\n|   └── grandchild2\n└── child2"
        );
    }

    #[test]
    fn test_deep_nesting() {
        let mut formatter = TreeFormatter::new();
        formatter.add_line(0, "root");
        formatter.add_line(1, "level1");
        formatter.add_line(2, "level2");
        formatter.add_line(3, "level3");
        formatter.add_line(4, "level4");
        let result = formatter.to_string();
        assert_eq!(
            result,
            "root\n└── level1\n    └── level2\n        └── level3\n            └── level4"
        );
    }

    #[test]
    fn test_complex_tree() {
        let mut formatter = TreeFormatter::new();
        formatter.add_line(0, "root");
        formatter.add_line(1, "branch1");
        formatter.add_line(2, "leaf1");
        formatter.add_line(2, "leaf2");
        formatter.add_line(1, "branch2");
        formatter.add_line(2, "leaf3");
        formatter.add_line(1, "branch3");
        let result = formatter.to_string();
        assert_eq!(result, "root\n├── branch1\n|   ├── leaf1\n|   └── leaf2\n├── branch2\n|   └── leaf3\n└── branch3");
    }

    #[test]
    fn test_add_tree() {
        let mut main_tree = TreeFormatter::new();
        main_tree.add_line(0, "main");
        main_tree.add_line(1, "first");

        let mut subtree = TreeFormatter::new();
        subtree.add_line(0, "sub_root");
        subtree.add_line(1, "sub_child1");
        subtree.add_line(1, "sub_child2");

        main_tree.add_tree(1, &subtree);
        main_tree.add_line(1, "last");

        let result = main_tree.to_string();
        assert_eq!(
            result,
            "main\n├── first\n├── sub_root\n|   ├── sub_child1\n|   └── sub_child2\n└── last"
        );
    }
}
