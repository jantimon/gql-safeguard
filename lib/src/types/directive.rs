use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum DirectiveType {
    Catch,
    ThrowOnFieldError,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct SourceLocation {
    pub line: usize,
    pub column: usize,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Directive {
    pub directive_type: DirectiveType,
    pub location: SourceLocation,
}
