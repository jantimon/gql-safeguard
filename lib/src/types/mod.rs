pub mod directive;
pub mod graphql;
pub mod violation;
