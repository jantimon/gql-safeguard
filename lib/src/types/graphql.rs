use crate::types::directive::Directive;
use serde::{Deserialize, Serialize};
use std::path::PathBuf;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum GraphQLItem {
    Query(QueryOperation),
    Fragment(FragmentDefinition),
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct QueryOperation {
    pub name: String,
    pub fields: Vec<Field>,
    pub fragments: Vec<FragmentSpread>,
    pub directives: Vec<Directive>,
    pub file_path: PathBuf,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct FragmentDefinition {
    pub name: String,
    pub fields: Vec<Field>,
    pub fragments: Vec<FragmentSpread>,
    pub directives: Vec<Directive>,
    pub file_path: PathBuf,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Field {
    pub name: String,
    pub directives: Vec<Directive>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct FragmentSpread {
    pub name: String,
    pub directives: Vec<Directive>,
}
