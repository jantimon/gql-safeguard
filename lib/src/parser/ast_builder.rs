use crate::scanner::extractor::GraphQLString;
use crate::types::directive::{Directive, DirectiveType, SourceLocation};
use crate::types::graphql::{
    Field, FragmentDefinition, FragmentSpread, GraphQLItem, QueryOperation,
};
use anyhow::Result;
use graphql_parser::parse_query;
use graphql_parser::query::{
    Definition, Document as QueryDocument, OperationDefinition, Selection, SelectionSet,
};
use graphql_parser::schema::Document;

pub struct AstBuilder;

impl AstBuilder {
    pub fn new() -> Self {
        Self
    }

    pub fn build_from_graphql_string(
        &self,
        graphql_string: &GraphQLString,
    ) -> Result<Vec<GraphQLItem>> {
        let document: QueryDocument<String> = parse_query(&graphql_string.content)?;
        let mut items = Vec::new();

        for definition in document.definitions {
            match definition {
                Definition::Operation(op) => {
                    if let Some(query) = self.convert_operation(op, &graphql_string.file_path)? {
                        items.push(GraphQLItem::Query(query));
                    }
                }
                Definition::Fragment(frag) => {
                    let fragment = self.convert_fragment(frag, &graphql_string.file_path)?;
                    items.push(GraphQLItem::Fragment(fragment));
                }
            }
        }

        Ok(items)
    }

    fn convert_operation(
        &self,
        op: OperationDefinition<String>,
        file_path: &std::path::Path,
    ) -> Result<Option<QueryOperation>> {
        match op {
            OperationDefinition::Query(query) => {
                let name = query.name.unwrap_or_else(|| "AnonymousQuery".to_string());
                let directives = self.convert_directives(&query.directives);
                let (fields, fragments) = self.convert_selection_set(&query.selection_set);

                Ok(Some(QueryOperation {
                    name,
                    fields,
                    fragments,
                    directives,
                    file_path: file_path.to_path_buf(),
                }))
            }
            OperationDefinition::Mutation(_) | OperationDefinition::Subscription(_) => {
                // Skip mutations and subscriptions for now
                Ok(None)
            }
            OperationDefinition::SelectionSet(_) => {
                // Skip bare selection sets for now
                Ok(None)
            }
        }
    }

    fn convert_fragment(
        &self,
        frag: graphql_parser::query::FragmentDefinition<String>,
        file_path: &std::path::Path,
    ) -> Result<FragmentDefinition> {
        let directives = self.convert_directives(&frag.directives);
        let (fields, fragments) = self.convert_selection_set(&frag.selection_set);

        Ok(FragmentDefinition {
            name: frag.name,
            fields,
            fragments,
            directives,
            file_path: file_path.to_path_buf(),
        })
    }

    fn convert_selection_set(
        &self,
        selection_set: &SelectionSet<String>,
    ) -> (Vec<Field>, Vec<FragmentSpread>) {
        let mut fields = Vec::new();
        let mut fragments = Vec::new();

        for selection in &selection_set.items {
            match selection {
                Selection::Field(field) => {
                    let directives = self.convert_directives(&field.directives);
                    fields.push(Field {
                        name: field.name.clone(),
                        directives,
                    });
                }
                Selection::FragmentSpread(spread) => {
                    let directives = self.convert_directives(&spread.directives);
                    fragments.push(FragmentSpread {
                        name: spread.fragment_name.clone(),
                        directives,
                    });
                }
                Selection::InlineFragment(_) => {
                    // Skip inline fragments for now
                }
            }
        }

        (fields, fragments)
    }

    fn convert_directives(
        &self,
        directives: &[graphql_parser::query::Directive<String>],
    ) -> Vec<Directive> {
        directives
            .iter()
            .filter_map(|dir| {
                let directive_type = match dir.name.as_str() {
                    "catch" => DirectiveType::Catch,
                    "throwOnFieldError" => DirectiveType::ThrowOnFieldError,
                    _ => return None,
                };

                Some(Directive {
                    directive_type,
                    location: SourceLocation { line: 1, column: 1 }, // Placeholder
                })
            })
            .collect()
    }
}
