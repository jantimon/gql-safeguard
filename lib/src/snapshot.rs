use crate::parser::ast_builder::AstBuilder;
use crate::scanner::{extractor::GraphQLExtractor, file_finder::FileFinder};
use crate::tree_formatter::TreeFormatter;
use crate::types::directive::DirectiveType;
use crate::types::graphql::{FragmentDefinition, GraphQLItem, QueryOperation};
use anyhow::Result;
use std::path::Path;

pub struct SnapshotGenerator {
    file_finder: FileFinder,
    extractor: GraphQLExtractor,
    ast_builder: AstBuilder,
}

impl SnapshotGenerator {
    pub fn new() -> Result<Self> {
        Ok(Self {
            file_finder: FileFinder::new("**/*.{ts,tsx}")?,
            extractor: GraphQLExtractor::new(),
            ast_builder: AstBuilder::new(),
        })
    }

    pub fn generate_snapshots_for_directory(&self, dir_path: &Path) -> Result<String> {
        let files = self.file_finder.find_files(dir_path)?;
        let mut snapshots = Vec::new();

        for file_path in files {
            let file_snapshot = self.generate_file_snapshot(&file_path)?;
            if !file_snapshot.is_empty() {
                snapshots.push(format!(
                    "=== {} ===\n{}",
                    file_path.display(),
                    file_snapshot
                ));
            }
        }

        Ok(snapshots.join("\n\n"))
    }

    fn generate_file_snapshot(&self, file_path: &Path) -> Result<String> {
        let graphql_strings = self.extractor.extract_from_file(file_path)?;
        if graphql_strings.is_empty() {
            return Ok(String::new());
        }

        let mut file_tree = TreeFormatter::new();
        file_tree.add_line(0, &format!("File: {}", file_path.display()));

        for (i, graphql_string) in graphql_strings.iter().enumerate() {
            let items = self.ast_builder.build_from_graphql_string(graphql_string)?;

            if !items.is_empty() {
                file_tree.add_line(1, &format!("GraphQL Block {}", i + 1));

                for item in items {
                    let item_tree = self.format_graphql_item(&item);
                    file_tree.add_tree(2, &item_tree);
                }
            }
        }

        Ok(file_tree.to_string())
    }

    fn format_graphql_item(&self, item: &GraphQLItem) -> TreeFormatter {
        match item {
            GraphQLItem::Query(query) => self.format_query(query),
            GraphQLItem::Fragment(fragment) => self.format_fragment(fragment),
        }
    }

    fn format_query(&self, query: &QueryOperation) -> TreeFormatter {
        let mut tree = TreeFormatter::new();

        let directives_str = self.format_directives_inline(&query.directives);
        tree.add_line(0, &format!("Query: {}{}", query.name, directives_str));

        // Add fields
        for field in &query.fields {
            let field_directives = self.format_directives_inline(&field.directives);
            tree.add_line(1, &format!("Field: {}{}", field.name, field_directives));
        }

        // Add fragment spreads
        for fragment in &query.fragments {
            let frag_directives = self.format_directives_inline(&fragment.directives);
            tree.add_line(
                1,
                &format!("Fragment: ...{}{}", fragment.name, frag_directives),
            );
        }

        tree
    }

    fn format_fragment(&self, fragment: &FragmentDefinition) -> TreeFormatter {
        let mut tree = TreeFormatter::new();

        let directives_str = self.format_directives_inline(&fragment.directives);
        tree.add_line(0, &format!("Fragment: {}{}", fragment.name, directives_str));

        // Add fields
        for field in &fragment.fields {
            let field_directives = self.format_directives_inline(&field.directives);
            tree.add_line(1, &format!("Field: {}{}", field.name, field_directives));
        }

        // Add fragment spreads
        for frag_spread in &fragment.fragments {
            let frag_directives = self.format_directives_inline(&frag_spread.directives);
            tree.add_line(
                1,
                &format!("Fragment: ...{}{}", frag_spread.name, frag_directives),
            );
        }

        tree
    }

    fn format_directives_inline(
        &self,
        directives: &[crate::types::directive::Directive],
    ) -> String {
        if directives.is_empty() {
            return String::new();
        }

        let directive_strs: Vec<String> = directives
            .iter()
            .map(|d| match d.directive_type {
                DirectiveType::Catch => "@catch".to_string(),
                DirectiveType::ThrowOnFieldError => "@throwOnFieldError ⚠️".to_string(),
            })
            .collect();

        format!(" {}", directive_strs.join(" "))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::scanner::extractor::GraphQLString;
    use std::path::PathBuf;

    #[test]
    fn test_simple_snapshot() {
        let ast_builder = AstBuilder::new();
        let generator = SnapshotGenerator::new().expect("Failed to create generator");

        // Create a simple GraphQL string to test parsing
        let test_graphql = GraphQLString {
            content: r#"
                query GetUser($id: ID!) @catch {
                    user(id: $id) {
                        id
                        name
                        avatar @throwOnFieldError
                    }
                }
            "#
            .to_string(),
            file_path: PathBuf::from("test.ts"),
            line: 1,
        };

        let items = ast_builder
            .build_from_graphql_string(&test_graphql)
            .expect("Failed to parse GraphQL");

        let mut snapshot_content = String::new();
        for item in items {
            let tree = generator.format_graphql_item(&item);
            snapshot_content.push_str(&tree.to_string());
            snapshot_content.push('\n');
        }

        insta::assert_snapshot!(snapshot_content.trim());
    }

    #[test]
    fn test_valid_fixtures_snapshot() {
        let generator = SnapshotGenerator::new().expect("Failed to create generator");
        let fixtures_dir = PathBuf::from("../fixtures/valid");

        if fixtures_dir.exists() {
            let snapshot = generator
                .generate_snapshots_for_directory(&fixtures_dir)
                .expect("Failed to generate snapshots");

            insta::assert_snapshot!("valid_fixtures", snapshot);

            // Basic assertions
            assert!(snapshot.contains("Query:") || snapshot.contains("Fragment:"));
        } else {
            panic!("Valid fixtures directory not found: {:?}", fixtures_dir);
        }
    }

    #[test]
    fn test_invalid_fixtures_snapshot() {
        let generator = SnapshotGenerator::new().expect("Failed to create generator");
        let fixtures_dir = PathBuf::from("../fixtures/invalid");

        if fixtures_dir.exists() {
            let snapshot = generator
                .generate_snapshots_for_directory(&fixtures_dir)
                .expect("Failed to generate snapshots");

            insta::assert_snapshot!("invalid_fixtures", snapshot);

            // Basic assertions
            assert!(snapshot.contains("Query:") || snapshot.contains("Fragment:"));
        } else {
            panic!("Invalid fixtures directory not found: {:?}", fixtures_dir);
        }
    }

    #[test]
    fn test_edge_cases_fixtures_snapshot() {
        let generator = SnapshotGenerator::new().expect("Failed to create generator");
        let fixtures_dir = PathBuf::from("../fixtures/edge_cases");

        if fixtures_dir.exists() {
            match generator.generate_snapshots_for_directory(&fixtures_dir) {
                Ok(snapshot) => {
                    insta::assert_snapshot!("edge_cases_fixtures", snapshot);
                }
                Err(e) => {
                    // Capture the error message as a snapshot too
                    let error_message = format!("Parsing error: {}", e);
                    insta::assert_snapshot!("edge_cases_error", error_message);
                }
            }
        } else {
            panic!(
                "Edge cases fixtures directory not found: {:?}",
                fixtures_dir
            );
        }
    }
}
