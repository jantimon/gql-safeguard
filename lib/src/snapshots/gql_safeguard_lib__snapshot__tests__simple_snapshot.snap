---
source: lib/src/snapshot.rs
expression: snapshot_content.trim()
---
Query: GetUser @catch
└── Field: user
